oss_client.cpp

```diff
@@ -0,0 +1,907 @@
+        0x9188,0x81a9,0xb1ca,0xa1eb,0xd10c,0xc12d,0xf14e,0xe16f,
+        0x1080,0x00a1,0x30c2,0x20e3,0x5004,0x4025,0x7046,0x6067,
+        0x83b9,0x9398,0xa3fb,0xb3da,0xc33d,0xd31c,0xe37f,0xf35e,
+        0x02b1,0x1290,0x22f3,0x32d2,0x4235,0x5214,0x6277,0x7256,
+        0xb5ea,0xa5cb,0x95a8,0x8589,0xf56e,0xe54f,0xd52c,0xc50d,
+        0x34e2,0x24c3,0x14a0,0x0481,0x7466,0x6447,0x5424,0x4405,
+        0xa7db,0xb7fa,0x8799,0x97b8,0xe75f,0xf77e,0xc71d,0xd73c,
+        0x26d3,0x36f2,0x0691,0x16b0,0x6657,0x7676,0x4615,0x5634,
+        0xd94c,0xc96d,0xf90e,0xe92f,0x99c8,0x89e9,0xb98a,0xa9ab,
+        0x5844,0x4865,0x7806,0x6827,0x18c0,0x08e1,0x3882,0x28a3,
+        0xcb7d,0xdb5c,0xeb3f,0xfb1e,0x8bf9,0x9bd8,0xabbb,0xbb9a,
+        0x4a75,0x5a54,0x6a37,0x7a16,0x0af1,0x1ad0,0x2ab3,0x3a92,
+        0xfd2e,0xed0f,0xdd6c,0xcd4d,0xbdaa,0xad8b,0x9de8,0x8dc9,
+        0x7c26,0x6c07,0x5c64,0x4c45,0x3ca2,0x2c83,0x1ce0,0x0cc1,
+        0xef1f,0xff3e,0xcf5d,0xdf7c,0xaf9b,0xbfba,0x8fd9,0x9ff8,
+        0x6e17,0x7e36,0x4e55,0x5e74,0x2e93,0x3eb2,0x0ed1,0x1ef0
+};
+
+static inline uint16_t crc16(const char *buf, size_t len) {
+    size_t counter;
+    uint16_t crc = 0;
+    for (counter = 0; counter < len; counter++)
+        crc = (crc<<8) ^ crc16tab[((crc>>8) ^ *buf++)&0x00FF];
+    return crc;
+}
+
+#define MAX_OSS_CLUSTER_HSLOT 16383
+#define MAIN_CONNECTION m_connections[0]
+
+static uint32_t calc_hslot_crc16_oss_cluster(const char *str, size_t length)
+{
+    uint32_t rv = (uint32_t) crc16(str, length) & MAX_OSS_CLUSTER_HSLOT;
+    return rv;
+}
+
+void oss_client_event_handler(evutil_socket_t sfd, short evtype, void *opaque)
+{
+    shard_connection *sc = (shard_connection *) opaque;
+
+    assert(sc != NULL);
+    assert(sc->sockfd == sfd);
+
+    sc->client->handle_event(evtype, sc);
+}
+
+shard_connection::shard_connection() : id(-1), client(NULL), sockfd(-1), read_buf(NULL), write_buf(NULL),
+                                       read_event(NULL), write_event(NULL), write_event_added(false), protocol(NULL),
+                                       pipeline(NULL), pending_reqs(0), pending_resp(0), min_slots_range(0), max_slots_range(0) {
+}
+
+shard_connection::~shard_connection() {
+    if (sockfd != -1) {
+        close(sockfd);
+        sockfd = -1;
+    }
+
+    if (read_buf != NULL) {
+        evbuffer_free(read_buf);
+        read_buf = NULL;
+    }
+
+    if (write_buf != NULL) {
+        evbuffer_free(write_buf);
+        write_buf = NULL;
+    }
+
+    if (read_event != NULL) {
+        event_free(read_event);
+        read_event = NULL;
+    }
+
+    if (write_event != NULL) {
+        event_free(write_event);
+        write_event = NULL;
+    }
+
+    if (protocol != NULL) {
+        delete protocol;
+        protocol = NULL;
+    }
+
+    if (pipeline != NULL) {
+        delete pipeline;
+        pipeline = NULL;
+    }
+}
+
+void shard_connection::setup_connection(int conn_id, oss_client* parent_client, abstract_protocol* abs_protocol) {
```

Comment:
There is quite a lot of duplicate code between oss_client.c and client.c, I think it should be avoided.